plugins/chanbackup.c

```diff
@@ -0,0 +1,304 @@
+#include "config.h"
+#include <ccan/array_size/array_size.h>
+#include <common/json_tok.h>
+#include <common/json.h>
+#include <common/json_stream.h>
+#include <ccan/tal/str/str.h>
+#include <plugins/libplugin.h>
+#include <ccan/read_write_all/read_write_all.h>
+#include <ccan/noerr/noerr.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <ccan/json_out/json_out.h>
+#include <ccan/cast/cast.h>
+#include <ccan/json_out/json_out.h>
+#include <ccan/mem/mem.h>
+#include <ccan/str/hex/hex.h>
+#include <plugins/libplugin.h>
+#include <common/type_to_string.h>
+#include <unistd.h>
+#include <sodium.h>
+#include <common/hsm_encryption.h>
+
+#include <ccan/err/err.h>
+#include <ccan/fdpass/fdpass.h>
+
+/* Global secret object to keep the derived encryption key for the SCB */
+struct secret secret;
+
+/* Helper to fetch out SCB from the RPC call */
+static bool json_to_str(const char *buffer, const jsmntok_t *tok,
+			const char **t)
+{
+	*t = json_strdup(tmpctx, buffer, tok);
+	return true;
+}
+
+/* This writes encrypted scb in the file */
+static void write_hsm(struct plugin *p, int fd, const jsmntok_t *scb_arr, 
+						const char *buf){
+
+	int i = 0;
+	const jsmntok_t *t;
+	char *data = tal(scb_arr, char);
+	
+	plugin_log(p,LOG_DBG, "buf is -----------------------------%s------------------------------------------", buf);
+
+	json_for_each_arr(i,t,scb_arr){
+		char *c = json_strdup(tmpctx, buf, json_get_arr(scb_arr, i));
+		
+		strcat(c, ",");
+
+		data = tal_strcat(scb_arr, data, c);
+	}
+	plugin_log(p,LOG_DBG, "data is -----------------------------%s------------------------------------------", data);
+	u8 *point = tal_dup_arr(scb_arr, u8, (u8 *)data, strlen(data), 0);
+	u8 *final = tal_arr(scb_arr, 
+						u8, 
+						tal_bytelen(point) + 
+							crypto_secretstream_xchacha20poly1305_ABYTES + 
+							crypto_secretstream_xchacha20poly1305_HEADERBYTES
+						);
+
+	crypto_secretstream_xchacha20poly1305_state crypto_state;
+
+	if (crypto_secretstream_xchacha20poly1305_init_push(&crypto_state, final,
+								(&secret)->data) != 0)
+		return;
+	
+	if (crypto_secretstream_xchacha20poly1305_push(
+							   &crypto_state,
+						       final + crypto_secretstream_xchacha20poly1305_HEADERBYTES,
+						       NULL, point,
+						       tal_bytelen(point),
+						       /* Additional data and tag */
+						       NULL, 0, 0))
+		return;
+	
+	if (!write_all(fd, final, tal_bytelen(final))) {
+			unlink_noerr("scb");
+	}
+}
+
+/* checks if the SCB file exists, creates a new one in case it doesn't. */
+static void maybe_create_new_scb(struct plugin *p, const char *scb_buf)
```

Comment:
This overwrites the file in place, which does not guarantee atomic operations. It's better to create a temporary file `recover.tmp`, write to that and then rename `recover.tmp` to `recover`. That guarantees that a partial write (plugin crash, loss of power, etc) doesn't corrupt your recovery file.

Reply:
Required reading: [UPDATE IN PLACE: A Poison Apple?](https://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.108.7810)